src/wild_encounter.c

```diff
@@ -0,0 +1,748 @@
+#include "global.h"
+#include "random.h"
+#include "wild_encounter.h"
+#include "event_data.h"
+#include "fieldmap.h"
+#include "roamer.h"
+#include "field_player_avatar.h"
+#include "battle_setup.h"
+#include "overworld.h"
+#include "metatile_behavior.h"
+#include "event_scripts.h"
+#include "script.h"
+#include "link.h"
+#include "quest_log.h"
+#include "constants/species.h"
+#include "constants/maps.h"
+#include "constants/vars.h"
+#include "constants/abilities.h"
+#include "constants/items.h"
+
+struct WildEncounterData
+{
+    u32 rngState;
+    u16 prevMetaTileBehavior;
+    u16 encounterRateBuff;
+    u8 stepsSinceLastEncounter;
+    u8 abilityEffect;
+    u16 leadMonHeldItem;
+};
+
+static EWRAM_DATA struct WildEncounterData sWildEncounterData = {};
+static EWRAM_DATA bool8 sWildEncountersDisabled = FALSE;
+
+extern const u8 gUnknown_83CA71C[][12];
+
+static bool8 UnlockedTanobyOrAreNotInTanoby(void);
+static u32 GenerateUnownPersonalityByLetter(u8 letter);
+static bool8 IsWildLevelAllowedByRepel(u8 level);
+static void ApplyFluteEncounterRateMod(u32 *rate);
+static u8 GetFluteEncounterRateModType(void);
+static void ApplyCleanseTagEncounterRateMod(u32 *rate);
+static u8 IsLeadMonHoldingCleanseTag(void);
+static u16 WildEncounterRandom(void);
+static void AddToWildEncounterRateBuff(u8 encouterRate);
+
+void DisableWildEncounters(bool8 state)
+{
+    sWildEncountersDisabled = state;
+}
+
+static u8 ChooseWildMonIndex_Land(void)
+{
+    u8 pct = Random() % 100;
+    if (pct < 20)
+        return 0;
+    if (pct >= 20 && pct < 40)
+        return 1;
+    if (pct >= 40 && pct < 50)
+        return 2;
+    if (pct >= 50 && pct < 60)
+        return 3;
+    if (pct >= 60 && pct < 70)
+        return 4;
+    if (pct >= 70 && pct < 80)
+        return 5;
+    if (pct >= 80 && pct < 85)
+        return 6;
+    if (pct >= 85 && pct < 90)
+        return 7;
+    if (pct >= 90 && pct < 94)
+        return 8;
+    if (pct >= 94 && pct < 98)
+        return 9;
+    if (pct == 98)
+        return 10;
+    return 11;
+}
+
+static u8 ChooseWildMonIndex_WaterRock(void)
+{
+    u8 pct = Random() % 100;
+    if (pct < 60)
+        return 0;
+    if (pct >= 60 && pct < 90)
+        return 1;
+    if (pct >= 90 && pct < 95)
+        return 2;
+    if (pct >= 95 && pct < 99)
+        return 3;
+    return 4;
+}
+
+static u8 ChooseWildMonIndex_Fishing(u8 rod)
+{
+    u8 slot = 0;
+    u8 pct = Random() % 100;
+    switch (rod)
+    {
+    case 0: // old
```

Comment:
Is it better to update case 1 and case 2 with their meanings?

Reply:
I'll do you one better.